components/usage/pkg/apiv1/usage.go

```diff
@@ -121,6 +122,116 @@ func (s *UsageService) ListBilledUsage(ctx context.Context, in *v1.ListBilledUsa
 	}, nil
 }
 
+func (s *UsageService) ListUsage(ctx context.Context, in *v1.ListUsageRequest) (*v1.ListUsageResponse, error) {
+	to := time.Now()
+	if in.To != nil {
+		to = in.To.AsTime()
+	}
+	from := to.Add(-maxQuerySize)
+	if in.From != nil {
+		from = in.From.AsTime()
+	}
+
+	if from.After(to) {
+		return nil, status.Errorf(codes.InvalidArgument, "Specified From timestamp is after To. Please ensure From is always before To")
+	}
+
+	if to.Sub(from) > maxQuerySize {
+		return nil, status.Errorf(codes.InvalidArgument, "Maximum range exceeded. Range specified can be at most %s", maxQuerySize.String())
+	}
+
+	order := db.DescendingOrder
+	if in.Order == v1.ListUsageRequest_ORDERING_ASCENDING {
+		order = db.AscendingOrder
+	}
+
+	var limit int64 = 1000
+	var page int64 = 0
+	var offset int64 = 0
+	if in.Pagination != nil {
+		limit = in.Pagination.PerPage
+		if limit < 1 {
+			return nil, status.Errorf(codes.InvalidArgument, "Specify a positive limit (was %d).", limit)
+		}
+		page = in.Pagination.Page
+		if page < 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "Specify a positive page number (was %d).", page)
+		}
+		offset = limit * (int64(math.Max(0, float64(page-1))))
+	}
+
+	listUsageResult, err := db.FindUsage(ctx, s.conn, &db.FindUsageParams{
+		AttributionId: db.AttributionID(in.GetAttributionId()),
+		From:          from,
+		To:            to,
+		Order:         order,
+		Offset:        offset,
+		Limit:         limit,
+	})
+	logger := log.Log.
+		WithField("attribution_id", in.AttributionId).
+		WithField("perPage", limit).
+		WithField("page", page).
+		WithField("from", from).
+		WithField("to", to)
+	if err != nil {
+		logger.WithError(err).Error("Failed to fetch usage.")
+		return nil, status.Error(codes.Internal, "unable to retrieve usage")
+	}
+
+	var usageData []*v1.Usage
+	for _, usageRecord := range listUsageResult {
+		kind := v1.Usage_KIND_WORKSPACE_INSTANCE
+		if usageRecord.Kind == db.InvoiceUsageKind {
+			kind = v1.Usage_KIND_INVOICE
+		}
+		attributionId, err := db.ParseAttributionID(string(usageRecord.AttributionID))
+		if err != nil {
+			logger.WithError(err).Errorf("Ignoring usage entry (ID: %s) with invalid attributionID: %s.", usageRecord.ID, usageRecord.AttributionID)
+			continue
+		}
```

Comment:
Persoanlly, I'd return an error here from the RPC. Yes, it would cause it to fail but it would more acutely show up and give us a chance to fix. There's a tendency that these logs are missed and never properly fixed.

I think I may have confused you in the previous comment on this. I think when we take the attribution ID from the DB, we can trust it to be the right shape so doing just `db.AttributionID(usageRecord.AttributionID)` should be fine. But I think we can do better. We can actually type the UsageRecord as an AttributionID directly, this removes this step.

But we should try to parse the Attribution ID from the incoming request. We currently use it on line 208 without parsing (and therefore at least some validation).

Reply:
Yes, I got indeed confused (somehow the comments show up on the wrong lines in VS Code sometimes). Makes much more sense now.